lib/ecmascript.ts

```diff
@@ -4826,91 +4858,111 @@ export function RoundDuration(
         ({ relativeTo, days: oneWeekDays } = MoveRelativeDate(calendar, relativeTo, oneWeek));
       }
       oneWeekDays = MathAbs(oneWeekDays);
-      const divisor = bigInt(oneWeekDays).multiply(dayLengthNs);
-      nanoseconds = divisor.multiply(weeks).plus(bigInt(days).multiply(dayLengthNs)).plus(nanoseconds);
-      const rounded = RoundNumberToIncrement(nanoseconds, divisor.multiply(increment).toJSNumber(), roundingMode);
-      total = nanoseconds.toJSNumber() / divisor.toJSNumber();
-      weeks = rounded.divide(divisor).toJSNumber();
-      nanoseconds = bigInt.zero;
+      const divisor = JSBI.multiply(JSBI.BigInt(oneWeekDays), dayLengthNs);
+      nanoseconds = JSBI.add(
+        JSBI.add(JSBI.multiply(divisor, JSBI.BigInt(weeks)), JSBI.multiply(JSBI.BigInt(days), dayLengthNs)),
+        nanoseconds
+      );
+      const rounded = RoundNumberToIncrement(
+        nanoseconds,
+        JSBI.toNumber(JSBI.multiply(divisor, JSBI.BigInt(increment))),
+        roundingMode
+      );
+      total = JSBI.toNumber(nanoseconds) / JSBI.toNumber(divisor);
+      weeks = JSBI.toNumber(JSBI.divide(rounded, divisor));
+      nanoseconds = ZERO;
       days = 0;
       break;
     }
     case 'day': {
-      const divisor = bigInt(dayLengthNs);
-      nanoseconds = divisor.multiply(days).plus(nanoseconds);
-      const rounded = RoundNumberToIncrement(nanoseconds, divisor.multiply(increment).toJSNumber(), roundingMode);
-      total = nanoseconds.toJSNumber() / divisor.toJSNumber();
-      days = rounded.divide(divisor).toJSNumber();
-      nanoseconds = bigInt.zero;
+      const divisor = dayLengthNs;
+      nanoseconds = JSBI.add(JSBI.multiply(divisor, JSBI.BigInt(days)), nanoseconds);
+      const rounded = RoundNumberToIncrement(
+        nanoseconds,
+        JSBI.toNumber(JSBI.multiply(divisor, JSBI.BigInt(increment))),
+        roundingMode
+      );
+      total = JSBI.toNumber(nanoseconds) / JSBI.toNumber(divisor);
+      days = JSBI.toNumber(JSBI.divide(rounded, divisor));
+      nanoseconds = ZERO;
       break;
     }
     case 'hour': {
       const divisor = 3600e9;
-      nanoseconds = bigInt(hours)
-        .multiply(3600e9)
-        .plus(bigInt(minutes).multiply(60e9))
-        .plus(bigInt(seconds).multiply(1e9))
-        .plus(bigInt(milliseconds).multiply(1e6))
-        .plus(bigInt(microseconds).multiply(1e3))
-        .plus(nanoseconds);
-      total = nanoseconds.toJSNumber() / divisor;
-      const rounded = RoundNumberToIncrement(nanoseconds, divisor * increment, roundingMode);
-      hours = rounded.divide(divisor).toJSNumber();
-      nanoseconds = bigInt.zero;
+      const hoursAsNanos = JSBI.multiply(JSBI.BigInt(hours), JSBI.BigInt(3600e9));
```

Comment:
Personally, I find this rewrite harder to read than the code it's replacing, and also pretty divergent from proposal-temporal.  What about making `allNanoseconds` a mutable value instead?  

Then you could approximate the current chained flow, e.g. 

```js
let allNanoseconds = JSBI.multiply(JSBI.BigInt(hours), JSBI.BigInt(3600e9));
allNanoseconds = JSBI.add(allNanoseconds, JSBI.multiply(JSBI.BigInt(minutes), JSBI.BigInt(60e9)));
. . .
allNanoseconds = JSBI.add(allNanoseconds, nanoseconds);
```

Reply:
I agree your suggestion looks a lot nicer. Done, both here and for the other cases.